File: 1840-maximum-building-height/solution.cpp
#include <vector>
#include <algorithm>
#include <cmath>

class Solution {
public:
    int maxBuilding(int n, std::vector<std::vector<int>>& restrictions) {
        restrictions.push_back({1, 0});
        std::sort(restrictions.begin(), restrictions.end());
        
        int m = restrictions.size();
        
        // Left to right pass: maintain slope constraint
        for (int i = 1; i < m; ++i) {
            restrictions[i][1] = std::min(restrictions[i][1], 
                                          restrictions[i - 1][1] + (restrictions[i][0] - restrictions[i - 1][0]));
        }
        
        // Right to left pass: maintain slope constraint
        for (int i = m - 2; i >= 0; --i) {
            restrictions[i][1] = std::min(restrictions[i][1], 
                                          restrictions[i + 1][1] + (restrictions[i + 1][0] - restrictions[i][0]));
        }
        
        int maxHeight = 0;
        // Calculate max height between adjacent restrictions
        for (int i = 0; i < m - 1; ++i) {
            int d = restrictions[i + 1][0] - restrictions[i][0];
            int h1 = restrictions[i][1];
            int h2 = restrictions[i + 1][1];
            // Peak between i and i+1: (h1 + h2 + dist) / 2
            maxHeight = std::max(maxHeight, (h1 + h2 + d) / 2);
        }
        
        // Check the range from the last restriction to building n
        maxHeight = std::max(maxHeight, restrictions.back()[1] + (n - restrictions.back()[0]));
        
        return maxHeight;
    }
};

File: 1840-maximum-building-height/solution.py
class Solution:
    def maxBuilding(self, n: int, restrictions: list[list[int]]) -> int:
        restrictions.append([1, 0])
        restrictions.sort()
        
        m = len(restrictions)
        
        # Left to right pass: ensure h[i] <= h[i-1] + dist
        for i in range(1, m):
            restrictions[i][1] = min(restrictions[i][1], 
                                     restrictions[i-1][1] + (restrictions[i][0] - restrictions[i-1][0]))
            
        # Right to left pass: ensure h[i] <= h[i+1] + dist
        for i in range(m - 2, -1, -1):
            restrictions[i][1] = min(restrictions[i][1], 
                                     restrictions[i+1][1] + (restrictions[i+1][0] - restrictions[i][0]))
            
        max_h = 0
        # Calculate max height between adjacent tightened restrictions
        for i in range(m - 1):
            d = restrictions[i+1][0] - restrictions[i][0]
            h1, h2 = restrictions[i][1], restrictions[i+1][1]
            max_h = max(max_h, (h1 + h2 + d) // 2)
            
        # Check range after the last restriction to n
        max_h = max(max_h, restrictions[-1][1] + (n - restrictions[-1][0]))
        
        return max_h